File: python-benchmarks/benchmarks/runner.py
"""
Benchmark Runner
=================
Orchestrates all format benchmarks, collects results and system information.
"""

import os
import platform
from datetime import datetime, timezone


class BenchmarkRunner:
    """Runs benchmarks for all configured message formats."""

    def __init__(self, iterations: int = 1000, warmup: int = 100, formats: list = None):
        self.iterations = iterations
        self.warmup = warmup
        self.requested_formats = formats or [
            "json", "bson", "protobuf", "capnproto", "msgpack", "avro", "flatbuffers"
        ]
        self._benchmarks = self._load_benchmarks()

    def _load_benchmarks(self) -> dict:
        """Load available benchmark implementations."""
        benchmarks = {}

        # JSON is always available (standard library)
        from .json_benchmark import JsonBenchmark
        benchmarks["json"] = JsonBenchmark()

        # BSON (via pymongo)
        try:
            from .bson_benchmark import BsonBenchmark
            benchmarks["bson"] = BsonBenchmark()
        except ImportError:
            print("⚠  BSON not available (install: pip install pymongo)")

        # MessagePack
        try:
            from .msgpack_benchmark import MsgpackBenchmark
            benchmarks["msgpack"] = MsgpackBenchmark()
        except ImportError:
            print("⚠  MessagePack not available (install: pip install msgpack)")

        # Protobuf
        try:
            from .protobuf_benchmark import ProtobufBenchmark
            benchmarks["protobuf"] = ProtobufBenchmark()
        except ImportError as e:
            print(f"⚠  Protobuf not available ({e})")
            print("   Run first: python compile_schemas.py")
        except Exception as e:
            print(f"⚠  Protobuf error: {e}")

        # Cap'n Proto
        try:
            from .capnproto_benchmark import CapnProtoBenchmark
            benchmarks["capnproto"] = CapnProtoBenchmark()
        except ImportError as e:
            print(f"⚠  Cap'n Proto not available ({e})")
            print("   Install: sudo apt-get install capnproto libcapnp-dev && pip install pycapnp")
        except Exception as e:
            print(f"⚠  Cap'n Proto error: {e}")

        # Apache Avro
        try:
            from .avro_benchmark import AvroBenchmark
            benchmarks["avro"] = AvroBenchmark()
        except ImportError as e:
            print(f"⚠  Apache Avro not available ({e})")
            print("   Install: pip install fastavro")
        except Exception as e:
            print(f"⚠  Apache Avro error: {e}")

        # FlatBuffers
        try:
            from .flatbuffers_benchmark import FlatBuffersBenchmark
            benchmarks["flatbuffers"] = FlatBuffersBenchmark()
        except ImportError as e:
            print(f"⚠  FlatBuffers not available ({e})")
            print("   Install: pip install flatbuffers")
        except Exception as e:
            print(f"⚠  FlatBuffers error: {e}")

        return benchmarks

    def run_all(self, test_data: dict) -> dict:
        """
        Run all benchmarks for the provided test data.

        Args:
            test_data: dict with {size_label: data_dict} pairs

        Returns:
            dict with all benchmark results
        """
        results = []
        skipped = []

        for format_key in self.requested_formats:
            if format_key not in self._benchmarks:
                print(f"\n⏭  {format_key} skipped (not available)")
                skipped.append(format_key)
                continue

            benchmark = self._benchmarks[format_key]
            print(f"\n📊 Benchmarking {benchmark.format_name}...")

            for size_label, data in test_data.items():
                print(f"   Payload size: {size_label}...", end=" ", flush=True)
                try:
                    result = benchmark.run_benchmark(
                        data, self.iterations, self.warmup
                    )
                    result["payload_size_label"] = size_label
                    results.append(result)
                    size_bytes = result["serialized_size_bytes"]
                    mean_ms = result["serialize_time_ms"]["mean"]
                    print(f"✓ ({size_bytes} bytes, {mean_ms:.4f} ms avg)")
                except Exception as e:
                    print(f"✗ error: {e}")

        return {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "system_info": self._get_system_info(),
            "config": {
                "iterations": self.iterations,
                "warmup": self.warmup,
                "formats": self.requested_formats,
                "payload_sizes": list(test_data.keys()),
                "skipped_formats": skipped,
            },
            "results": results,
        }

    @staticmethod
    def _get_system_info() -> dict:
        """Collect system information."""
        return {
            "platform": platform.platform(),
            "python_version": platform.python_version(),
            "processor": BenchmarkRunner._get_cpu_name(),
            "machine": platform.machine(),
            "cpu_count": os.cpu_count() or 0,
        }

    @staticmethod
    def _get_cpu_name() -> str:
        """Read CPU name, with fallback for Linux/WSL/Docker."""
        # Try platform.processor() first
        name = platform.processor()
        if name and name != "unknown":
            return name

        # Fallback: read /proc/cpuinfo (Linux)
        try:
            with open("/proc/cpuinfo", "r") as f:
                for line in f:
                    if line.startswith("model name"):
                        return line.split(":", 1)[1].strip()
        except (FileNotFoundError, PermissionError):
            pass

        return "unknown"

File: go-benchmarks/benchmarks/runner.go
package benchmarks

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"time"

	"example.com/benchmarks/models"
)

// RunResult holds the complete result set from a benchmark run.
type RunResult struct {
	Timestamp  string        `json:"timestamp"`
	SystemInfo GoSystemInfo  `json:"system_info"`
	Config     RunConfig     `json:"config"`
	Results    []BenchmarkResult `json:"results"`
}

// GoSystemInfo holds system information with correct JSON types.
type GoSystemInfo struct {
	Platform      string `json:"platform"`
	GoVersion     string `json:"go_version"`
	PythonVersion string `json:"python_version,omitempty"`
	Processor     string `json:"processor"`
	Machine       string `json:"machine"`
	CpuCount      int    `json:"cpu_count"`
	Language      string `json:"language"`
}

// RunConfig holds the configuration used for this benchmark run.
type RunConfig struct {
	Iterations     int      `json:"iterations"`
	Warmup         int      `json:"warmup"`
	Formats        []string `json:"formats"`
	PayloadSizes   []string `json:"payload_sizes"`
	SkippedFormats []string `json:"skipped_formats"`
}

// Runner orchestrates all format benchmarks.
type Runner struct {
	Iterations int
	Warmup     int
	Formats    []string
	benchmarks map[string]FormatBenchmark
}

// NewRunner creates a new benchmark runner with the given configuration.
func NewRunner(iterations, warmup int, formats []string) *Runner {
	r := &Runner{
		Iterations: iterations,
		Warmup:     warmup,
		Formats:    formats,
		benchmarks: make(map[string]FormatBenchmark),
	}
	r.loadBenchmarks()
	return r
}

func (r *Runner) loadBenchmarks() {
	// JSON - always available (stdlib)
	r.benchmarks["json"] = &JSONBenchmark{}

	// BSON
	r.benchmarks["bson"] = &BSONBenchmark{}

	// MessagePack
	r.benchmarks["msgpack"] = &MsgpackBenchmark{}

	// Protobuf
	r.benchmarks["protobuf"] = &ProtobufBenchmark{}

	// Cap'n Proto
	r.benchmarks["capnproto"] = &CapnProtoBenchmark{}

	// Apache Avro
	avro, err := NewAvroBenchmark()
	if err != nil {
		fmt.Fprintf(os.Stderr, "⚠  Avro not available: %v\n", err)
	} else {
		r.benchmarks["avro"] = avro
	}

	// FlatBuffers
	r.benchmarks["flatbuffers"] = &FlatBuffersBenchmark{}
}

// dataToMap converts a BenchmarkMessage struct to map[string]interface{} for generic benchmark use.
func dataToMap(msg models.BenchmarkMessage) map[string]interface{} {
	m := map[string]interface{}{
		"id":        msg.ID,
		"timestamp": msg.Timestamp,
		"username":  msg.Username,
		"email":     msg.Email,
		"content":   msg.Content,
		"tags":      msg.Tags,
		"metadata":  msg.Metadata,
		"score":     msg.Score,
		"is_active": msg.IsActive,
		"items":     itemsToSlice(msg.Items),
	}
	if msg.NestedData != nil {
		m["nested_data"] = map[string]interface{}{
			"field1": msg.NestedData.Field1,
			"field2": msg.NestedData.Field2,
			"values": msg.NestedData.Values,
		}
	}
	return m
}

func itemsToSlice(items []models.Item) []interface{} {
	result := make([]interface{}, len(items))
	for i, item := range items {
		result[i] = map[string]interface{}{
			"name":        item.Name,
			"value":       item.Value,
			"active":      item.Active,
			"description": item.Description,
			"tags":        item.Tags,
		}
	}
	return result
}

// RunAll executes all benchmarks for the given test data.
func (r *Runner) RunAll(testData map[string]models.BenchmarkMessage) *RunResult {
	var results []BenchmarkResult
	var skipped []string

	for _, formatKey := range r.Formats {
		bench, ok := r.benchmarks[formatKey]
		if !ok {
			fmt.Printf("\n⏭  %s skipped (not available)\n", formatKey)
			skipped = append(skipped, formatKey)
			continue
		}

		fmt.Printf("\n📊 Benchmarking %s...\n", bench.FormatName())

		for sizeLabel, data := range testData {
			fmt.Printf("   Payload size: %s... ", sizeLabel)
			dataMap := dataToMap(data)

			result, err := RunBenchmark(bench, dataMap, r.Iterations, r.Warmup)
			if err != nil {
				fmt.Printf("✗ error: %v\n", err)
				continue
			}
			result.PayloadSizeLabel = sizeLabel
			results = append(results, *result)
			fmt.Printf("✓ (%d bytes, %.4f ms avg)\n",
				result.SerializedSize, result.SerializeTime.Mean)
		}
	}

	sizes := make([]string, 0, len(testData))
	for k := range testData {
		sizes = append(sizes, k)
	}

	return &RunResult{
		Timestamp:  time.Now().UTC().Format(time.RFC3339),
		SystemInfo: getSystemInfo(),
		Config: RunConfig{
			Iterations:     r.Iterations,
			Warmup:         r.Warmup,
			Formats:        r.Formats,
			PayloadSizes:   sizes,
			SkippedFormats: skipped,
		},
		Results: results,
	}
}

func getSystemInfo() GoSystemInfo {
	info := GoSystemInfo{
		Platform:  runtime.GOOS + "/" + runtime.GOARCH,
		GoVersion: runtime.Version(),
		Machine:   runtime.GOARCH,
		CpuCount:  runtime.NumCPU(),
		Language:  "go",
		Processor: "unknown",
	}

	// Try to read CPU name from /proc/cpuinfo (Linux/WSL)
	data, err := os.ReadFile("/proc/cpuinfo")
	if err == nil {
		for _, line := range strings.Split(string(data), "\n") {
			if strings.HasPrefix(line, "model name") {
				parts := strings.SplitN(line, ":", 2)
				if len(parts) == 2 {
					info.Processor = strings.TrimSpace(parts[1])
					break
				}
			}
		}
	}

	return info
}
